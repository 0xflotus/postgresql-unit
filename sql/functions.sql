-- text extractors
SELECT value('2'::unit);
SELECT value(meter(2));
SELECT dimension('2'::unit);
SELECT dimension(kilogram(2));

-- test unit addition/subtraction
SELECT '1'::unit + '2' AS sum;
SELECT '1'::unit - '2' AS difference;
SELECT '3 m'::unit - '1 Âµm' AS difference;
SELECT '0' + meter() AS error;
SELECT meter() + '0' AS error;
SELECT meter() + kilogram() AS error;
SELECT meter() - kilogram() AS error;

-- test unit multiplication/division
SELECT - '1'::unit AS negative;
SELECT '2'::unit * '2' AS product;
SELECT '5'::unit / '2' AS fraction;
SELECT '0'::unit / '2' AS zero;
SELECT '5'::unit / '0' AS division_by_zero;

SELECT - second() AS minus_second;
SELECT meter() * '2' AS two_meters;
SELECT '3' * meter() AS three_meters;
SELECT meter() * meter() AS square_meter;
SELECT meter() * kilogram() AS meter_kilogram;
SELECT '1' / second(.02) AS hertz;
SELECT meter(9.81) / (second() * second()) AS acceleration;

-- test unit/double operators
SELECT 2 * meter() AS two_meters;
SELECT ampere() * 3 AS three_ampere;
SELECT 4 / second() AS four_hertz;
SELECT kelvin(10) / 5 AS two_kelvin;
SELECT 0 / candela() AS zero_by_candela;
SELECT 1 / mole(0) AS error;
SELECT byte(0) / 2 AS zero_byte;
SELECT kilogram(1) / 0 AS error;

-- test exponentiation
SELECT '6'::unit ^ 2 AS square;
SELECT meter(100) ^ 2 AS square_meter;
SELECT kilogram(100) ^ 2 AS square_kilogram;
SELECT second(.02) ^ -1 AS hertz;
SELECT ampere(4) ^ 0 AS unity;
SELECT '0'::unit ^ 0 AS unity;

-- test roots
SELECT sqrt('4'::unit);
SELECT sqrt('4 m^2'::unit);
SELECT |/'16 s^-4'::unit;
SELECT sqrt('-4 m^2'::unit) AS error;
SELECT sqrt('4 m'::unit) AS error;
SELECT cbrt('8'::unit);
SELECT cbrt('8 m^3'::unit);
SELECT ||/'-27 s^-6'::unit;
SELECT cbrt('-4 m'::unit) AS error;

WITH
  v(u) AS (VALUES
	(NULL),
	('0'::unit),
	('1'::unit),
	('2'::unit),
	(meter(0)),
	(meter()),
	(meter() * '2'),
	(kilogram(0)),
	(kilogram()),
	(kilogram() * '2')
  ),
  va(a) AS (SELECT * FROM v),
  vb(b) AS (SELECT * FROM v)
SELECT
  a, b,
  -a AS neg,
  a * b AS mul,
  CASE WHEN value(b) = 0 THEN NULL ELSE a / b END AS div
FROM
  va CROSS JOIN vb;

WITH
  v(a) AS (VALUES
	(NULL),
	('1'::unit),
	('2'::unit),
	(meter()),
	(meter() * '2'),
	(kilogram()),
	(kilogram() * '2')
  )
SELECT
  a,
  a ^ -2 AS pow_2, a ^ -1 AS pow_1, a ^ 0 AS pow0, a ^ 1 AS pow1, a ^ 2 AS pow2
FROM
  v;

-- test conversion
SELECT '1m'::unit @ 'mm';
SELECT '10g'::unit @ 'mg';
SELECT '5dm^3'::unit @ 'l';
SELECT '9.81 m/s^2'::unit @ 'N/kg';
SELECT '16384 B'::unit @ '8192 B';
SELECT '1 hl'::unit @ '0.5 l';
SELECT '5MB/min'::unit @ 'GB/d';
SELECT '5m'::unit @ 's' AS error;
