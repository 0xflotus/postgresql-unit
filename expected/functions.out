-- test functions
SELECT value('2'::unit);
 value 
-------
     2
(1 row)

SELECT value(meter(2));
 value 
-------
     2
(1 row)

SELECT dimension('2'::unit);
 dimension 
-----------
 1
(1 row)

SELECT dimension(kilogram(2));
 dimension 
-----------
 1 kg
(1 row)

-- test unit addition/subtraction
SELECT '1'::unit + '2' AS sum;
 sum 
-----
 3
(1 row)

SELECT '1'::unit - '2' AS difference;
 difference 
------------
 -1
(1 row)

SELECT '3 m'::unit - '1 Âµm' AS difference;
 difference 
------------
 2.999999 m
(1 row)

SELECT '0' + meter() AS error;
ERROR:  dimension mismatch in "+" operation: "0", "1 m"
SELECT meter() + '0' AS error;
ERROR:  dimension mismatch in "+" operation: "1 m", "0"
SELECT meter() + kilogram() AS error;
ERROR:  dimension mismatch in "+" operation: "1 m", "1 kg"
SELECT meter() - kilogram() AS error;
ERROR:  dimension mismatch in "-" operation: "1 m", "1 kg"
-- test unit multiplication/division
SELECT - '1'::unit AS negative;
 negative 
----------
 -1
(1 row)

SELECT '2'::unit * '2' AS product;
 product 
---------
 4
(1 row)

SELECT '5'::unit / '2' AS fraction;
 fraction 
----------
 2.5
(1 row)

SELECT '0'::unit / '2' AS zero;
 zero 
------
 0
(1 row)

SELECT '5'::unit / '0' AS division_by_zero;
ERROR:  division by zero-valued unit: "0"
SELECT - second() AS minus_second;
 minus_second 
--------------
 -1 s
(1 row)

SELECT meter() * '2' AS two_meters;
 two_meters 
------------
 2 m
(1 row)

SELECT '3' * meter() AS three_meters;
 three_meters 
--------------
 3 m
(1 row)

SELECT meter() * meter() AS square_meter;
 square_meter 
--------------
 1 m^2
(1 row)

SELECT meter() * kilogram() AS meter_kilogram;
 meter_kilogram 
----------------
 1 m*kg
(1 row)

SELECT '1' / second(.02) AS hertz;
 hertz 
-------
 50 Hz
(1 row)

SELECT meter(9.81) / (second() * second()) AS acceleration;
 acceleration 
--------------
 9.81 m/s^2
(1 row)

-- test unit/double operators
SELECT 2 * meter() AS two_meters;
 two_meters 
------------
 2 m
(1 row)

SELECT ampere() * 3 AS three_ampere;
 three_ampere 
--------------
 3 A
(1 row)

SELECT 4 / second() AS four_hertz;
 four_hertz 
------------
 4 Hz
(1 row)

SELECT kelvin(10) / 5 AS two_kelvin;
 two_kelvin 
------------
 2 K
(1 row)

SELECT 0 / candela() AS zero_by_candela;
 zero_by_candela 
-----------------
 0 cd^-1
(1 row)

SELECT 1 / mole(0) AS error;
ERROR:  division by zero-valued unit: "0 mol"
SELECT byte(0) / 2 AS zero_byte;
 zero_byte 
-----------
 0 B
(1 row)

SELECT kilogram(1) / 0 AS error;
ERROR:  division of unit by zero
-- test exponentiation
SELECT '6'::unit ^ 2 AS square;
 square 
--------
 36
(1 row)

SELECT meter(100) ^ 2 AS square_meter;
 square_meter 
--------------
 10000 m^2
(1 row)

SELECT kilogram(100) ^ 2 AS square_kilogram;
 square_kilogram 
-----------------
 10000 kg^2
(1 row)

SELECT second(.02) ^ -1 AS hertz;
 hertz 
-------
 50 Hz
(1 row)

SELECT ampere(4) ^ 0 AS unity;
 unity 
-------
 1
(1 row)

SELECT '0'::unit ^ 0 AS unity;
 unity 
-------
 1
(1 row)

-- test roots
SELECT sqrt('4'::unit);
 sqrt 
------
 2
(1 row)

SELECT sqrt('4 m^2'::unit);
 sqrt 
------
 2 m
(1 row)

SELECT |/'16 s^-4'::unit;
 ?column? 
----------
 4 s^-2
(1 row)

SELECT sqrt('-4 m^2'::unit) AS error;
ERROR:  cannot take square root of a negative-valued unit
SELECT sqrt('4 m'::unit) AS error;
ERROR:  cannot take square root of a unit with odd "m" exponent
SELECT cbrt('8'::unit);
 cbrt 
------
 2
(1 row)

SELECT cbrt('8 m^3'::unit);
 cbrt 
------
 2 m
(1 row)

SELECT ||/'-27 s^-6'::unit;
 ?column? 
----------
 -3 s^-2
(1 row)

SELECT cbrt('-4 m'::unit) AS error;
ERROR:  cannot take cube root of a unit with "m" exponent not divisible by three
WITH
  v(u) AS (VALUES
	(NULL),
	('0'::unit),
	('1'::unit),
	('2'::unit),
	(meter(0)),
	(meter()),
	(meter() * '2'),
	(kilogram(0)),
	(kilogram()),
	(kilogram() * '2')
  ),
  va(a) AS (SELECT * FROM v),
  vb(b) AS (SELECT * FROM v)
SELECT
  a, b,
  -a AS neg,
  a * b AS mul,
  CASE WHEN value(b) = 0 THEN NULL ELSE a / b END AS div
FROM
  va CROSS JOIN vb;
  a   |  b   |  neg  |  mul   |    div    
------+------+-------+--------+-----------
      |      |       |        | 
      | 0    |       |        | 
      | 1    |       |        | 
      | 2    |       |        | 
      | 0 m  |       |        | 
      | 1 m  |       |        | 
      | 2 m  |       |        | 
      | 0 kg |       |        | 
      | 1 kg |       |        | 
      | 2 kg |       |        | 
 0    |      | -0    |        | 
 0    | 0    | -0    | 0      | 
 0    | 1    | -0    | 0      | 0
 0    | 2    | -0    | 0      | 0
 0    | 0 m  | -0    | 0 m    | 
 0    | 1 m  | -0    | 0 m    | 0 m^-1
 0    | 2 m  | -0    | 0 m    | 0 m^-1
 0    | 0 kg | -0    | 0 kg   | 
 0    | 1 kg | -0    | 0 kg   | 0 kg^-1
 0    | 2 kg | -0    | 0 kg   | 0 kg^-1
 1    |      | -1    |        | 
 1    | 0    | -1    | 0      | 
 1    | 1    | -1    | 1      | 1
 1    | 2    | -1    | 2      | 0.5
 1    | 0 m  | -1    | 0 m    | 
 1    | 1 m  | -1    | 1 m    | 1 m^-1
 1    | 2 m  | -1    | 2 m    | 0.5 m^-1
 1    | 0 kg | -1    | 0 kg   | 
 1    | 1 kg | -1    | 1 kg   | 1 kg^-1
 1    | 2 kg | -1    | 2 kg   | 0.5 kg^-1
 2    |      | -2    |        | 
 2    | 0    | -2    | 0      | 
 2    | 1    | -2    | 2      | 2
 2    | 2    | -2    | 4      | 1
 2    | 0 m  | -2    | 0 m    | 
 2    | 1 m  | -2    | 2 m    | 2 m^-1
 2    | 2 m  | -2    | 4 m    | 1 m^-1
 2    | 0 kg | -2    | 0 kg   | 
 2    | 1 kg | -2    | 2 kg   | 2 kg^-1
 2    | 2 kg | -2    | 4 kg   | 1 kg^-1
 0 m  |      | -0 m  |        | 
 0 m  | 0    | -0 m  | 0 m    | 
 0 m  | 1    | -0 m  | 0 m    | 0 m
 0 m  | 2    | -0 m  | 0 m    | 0 m
 0 m  | 0 m  | -0 m  | 0 m^2  | 
 0 m  | 1 m  | -0 m  | 0 m^2  | 0
 0 m  | 2 m  | -0 m  | 0 m^2  | 0
 0 m  | 0 kg | -0 m  | 0 m*kg | 
 0 m  | 1 kg | -0 m  | 0 m*kg | 0 m/kg
 0 m  | 2 kg | -0 m  | 0 m*kg | 0 m/kg
 1 m  |      | -1 m  |        | 
 1 m  | 0    | -1 m  | 0 m    | 
 1 m  | 1    | -1 m  | 1 m    | 1 m
 1 m  | 2    | -1 m  | 2 m    | 500 mm
 1 m  | 0 m  | -1 m  | 0 m^2  | 
 1 m  | 1 m  | -1 m  | 1 m^2  | 1
 1 m  | 2 m  | -1 m  | 2 m^2  | 0.5
 1 m  | 0 kg | -1 m  | 0 m*kg | 
 1 m  | 1 kg | -1 m  | 1 m*kg | 1 m/kg
 1 m  | 2 kg | -1 m  | 2 m*kg | 500 mm/kg
 2 m  |      | -2 m  |        | 
 2 m  | 0    | -2 m  | 0 m    | 
 2 m  | 1    | -2 m  | 2 m    | 2 m
 2 m  | 2    | -2 m  | 4 m    | 1 m
 2 m  | 0 m  | -2 m  | 0 m^2  | 
 2 m  | 1 m  | -2 m  | 2 m^2  | 2
 2 m  | 2 m  | -2 m  | 4 m^2  | 1
 2 m  | 0 kg | -2 m  | 0 m*kg | 
 2 m  | 1 kg | -2 m  | 2 m*kg | 2 m/kg
 2 m  | 2 kg | -2 m  | 4 m*kg | 1 m/kg
 0 kg |      | -0 kg |        | 
 0 kg | 0    | -0 kg | 0 kg   | 
 0 kg | 1    | -0 kg | 0 kg   | 0 kg
 0 kg | 2    | -0 kg | 0 kg   | 0 kg
 0 kg | 0 m  | -0 kg | 0 m*kg | 
 0 kg | 1 m  | -0 kg | 0 m*kg | 0 kg/m
 0 kg | 2 m  | -0 kg | 0 m*kg | 0 kg/m
 0 kg | 0 kg | -0 kg | 0 kg^2 | 
 0 kg | 1 kg | -0 kg | 0 kg^2 | 0
 0 kg | 2 kg | -0 kg | 0 kg^2 | 0
 1 kg |      | -1 kg |        | 
 1 kg | 0    | -1 kg | 0 kg   | 
 1 kg | 1    | -1 kg | 1 kg   | 1 kg
 1 kg | 2    | -1 kg | 2 kg   | 500 g
 1 kg | 0 m  | -1 kg | 0 m*kg | 
 1 kg | 1 m  | -1 kg | 1 m*kg | 1 kg/m
 1 kg | 2 m  | -1 kg | 2 m*kg | 500 g/m
 1 kg | 0 kg | -1 kg | 0 kg^2 | 
 1 kg | 1 kg | -1 kg | 1 kg^2 | 1
 1 kg | 2 kg | -1 kg | 2 kg^2 | 0.5
 2 kg |      | -2 kg |        | 
 2 kg | 0    | -2 kg | 0 kg   | 
 2 kg | 1    | -2 kg | 2 kg   | 2 kg
 2 kg | 2    | -2 kg | 4 kg   | 1 kg
 2 kg | 0 m  | -2 kg | 0 m*kg | 
 2 kg | 1 m  | -2 kg | 2 m*kg | 2 kg/m
 2 kg | 2 m  | -2 kg | 4 m*kg | 1 kg/m
 2 kg | 0 kg | -2 kg | 0 kg^2 | 
 2 kg | 1 kg | -2 kg | 2 kg^2 | 2
 2 kg | 2 kg | -2 kg | 4 kg^2 | 1
(100 rows)

WITH
  v(a) AS (VALUES
	(NULL),
	('1'::unit),
	('2'::unit),
	(meter()),
	(meter() * '2'),
	(kilogram()),
	(kilogram() * '2')
  )
SELECT
  a,
  a ^ -2 AS pow_2, a ^ -1 AS pow_1, a ^ 0 AS pow0, a ^ 1 AS pow1, a ^ 2 AS pow2
FROM
  v;
  a   |   pow_2    |   pow_1   | pow0 | pow1 |  pow2  
------+------------+-----------+------+------+--------
      |            |           |      |      | 
 1    | 1          | 1         | 1    | 1    | 1
 2    | 0.25       | 0.5       | 1    | 2    | 4
 1 m  | 1 m^-2     | 1 m^-1    | 1    | 1 m  | 1 m^2
 2 m  | 0.25 m^-2  | 0.5 m^-1  | 1    | 2 m  | 4 m^2
 1 kg | 1 kg^-2    | 1 kg^-1   | 1    | 1 kg | 1 kg^2
 2 kg | 0.25 kg^-2 | 0.5 kg^-1 | 1    | 2 kg | 4 kg^2
(7 rows)

-- test conversion
SELECT '1m'::unit @ 'mm';
 ?column? 
----------
 1000 mm
(1 row)

SELECT '10g'::unit @ 'mg';
 ?column? 
----------
 10000 mg
(1 row)

SELECT '5dm^3'::unit @ 'l';
 ?column? 
----------
 5 l
(1 row)

SELECT '9.81 m/s^2'::unit @ 'N/kg', '9.81 m/s^2'::unit @@ 'N/kg';
 ?column?  | ?column? 
-----------+----------
 9.81 N/kg |     9.81
(1 row)

SELECT '16384 B'::unit @ '8192 B', '16384 B'::unit @@ '8192 B';
  ?column?  | ?column? 
------------+----------
 2 * 8192 B |        2
(1 row)

SELECT '1 hl'::unit @ '0.5 l';
  ?column?   
-------------
 200 * 0.5 l
(1 row)

SELECT '5MB/min'::unit @ 'GB/d';
 ?column? 
----------
 7.2 GB/d
(1 row)

SELECT '5m'::unit @ 's' AS error;
ERROR:  dimension mismatch in "@" operation: "5 m", "1 s"
